tools/src/main/java/org/apache/kafka/tools/GetOffsetShell.java

```diff
@@ -0,0 +1,404 @@
+        List<String> ruleSpecs = Arrays.asList(topicPartitions.split(","));
+        List<TopicPartitionFilter> rules = ruleSpecs.stream().map(ruleSpec -> {
+            try {
+                return parseRuleSpec(ruleSpec);
+            } catch (TerseException e) {
+                throw new RuntimeException(e);
+            }
+        }).collect(Collectors.toList());
+
+        return new CompositeTopicPartitionFilter(rules);
+    }
+
+    /**
+     * Creates a topic-partition filter based on a topic pattern and a set of partition ids.
+     */
+    public TopicPartitionFilter createTopicPartitionFilterWithTopicAndPartitionPattern(String topicOpt, String partitionIds) throws TerseException {
+        return new TopicFilterAndPartitionFilter(
+                new IncludeList(topicOpt != null ? topicOpt : ".*"),
+                new PartitionsSetFilter(createPartitionSet(partitionIds))
+        );
+    }
+
+    private Set<Integer> createPartitionSet(String partitionsString) throws TerseException {
+        Set<Integer> partitions;
+
+        if (partitionsString == null || partitionsString.isEmpty()) {
+            partitions = Collections.emptySet();
+        } else {
+            try {
+                partitions = Arrays.stream(partitionsString.split(",")).map(Integer::parseInt).collect(Collectors.toSet());
+            } catch (NumberFormatException e) {
+                throw new TerseException("--partitions expects a comma separated list of numeric " +
+                        "partition ids, but received: " + partitionsString);
+            }
+        }
+
+        return partitions;
+    }
+
+    /**
+     * Return the partition infos. Filter them with topicPartitionFilter.
+     */
+    private List<TopicPartition> listPartitionInfos(
+            Admin client,
+            TopicPartitionFilter topicPartitionFilter,
+            boolean excludeInternalTopics
+    ) throws ExecutionException, InterruptedException {
+        ListTopicsOptions listTopicsOptions = new ListTopicsOptions().listInternal(!excludeInternalTopics);
+        Set<String> topics = client.listTopics(listTopicsOptions).names().get();
+        Set<String> filteredTopics = topics.stream().filter(topicPartitionFilter::isTopicAllowed).collect(Collectors.toSet());
+
+        return client.describeTopics(filteredTopics).allTopicNames().get().entrySet().stream().flatMap(
+                topic -> topic.getValue().partitions().stream().map(
+                        tp -> new TopicPartition(topic.getKey(), tp.partition())
+                ).filter(topicPartitionFilter::isTopicPartitionAllowed)
+        ).collect(Collectors.toList());
+    }
+
+    private TopicPartitionFilter parseRuleSpec(String ruleSpec) throws TerseException, RuntimeException {
+        Matcher matcher = TOPIC_PARTITION_PATTERN.matcher(ruleSpec);
+
+        if (!matcher.matches())
+            throw new TerseException("Invalid rule specification: " + ruleSpec);
```

Comment:
We used to throw `IllegalArgumentException` why are we using `TerseException` now?

Reply:
It was a recommendation of @fvaleri

Reply:
... and there was [a reason for that](https://github.com/apache/kafka/pull/13562#pullrequestreview-1518389971). The old code was catching these exceptions and printing the error message. Instead, your code was printing the error message plus an ugly stack trace. That's why I suggested to use TerseException, as we have in other tools.

Reply:
Fair enough, thanks for the clarification